### ISSUE TITLE
Chapter 3 : Exercise 1 - MNIST Classifier with 97% accuracy - Could not pickle the task to send it to the workers.

### ISSUE THREAD

**Issue by a-d14** [ageron/handson-ml#658]
Ran this code for the first exercise - 
```
from sklearn.model_selection import GridSearchCV
param_grid = [{'weights': ["uniform", "distance"], 'n_neighbors': [3, 4, 5]}]
knn_clf = KNeighborsClassifier()
grid_search = GridSearchCV(knn_clf, param_grid, cv=5, verbose=3, n_jobs=-1)
grid_search.fit(X_train, y_train)
```

I got this error - 
Could not pickle the task to send it to the workers.

How do I resolve this?

**Comment by Rafayqayyum** [comment]
I was unable to reproduce this error. Try updating your sklearn and joblib.
`pip install --upgrade [Package name]`

or if you have conda installed
`conda update [package name]`

**Comment by GeoSegun** [comment]
This error usually occurs when you are trying to pass an object to the workers for parallel processing that cannot be pickled. Here, it seems like the KNeighborsClassifier() object is the one causing the issue.

One solution is to define a function that creates and returns the KNeighborsClassifier() object, and then pass this function to GridSearchCV() instead of the object itself.

`from sklearn.model_selection import GridSearchCV`
`from sklearn.neighbors import KNeighborsClassifier`

`def create_knn():`
    `return KNeighborsClassifier()`

`param_grid = [{'weights': ["uniform", "distance"], 'n_neighbors': [3, 4, 5]}]`

`knn_clf = create_knn()`

`grid_search = GridSearchCV(knn_clf, param_grid, cv=5, verbose=3, n_jobs=-1)`

`grid_search.fit(X_train, y_train)`

This should solve the error you are facing

**Comment by maciejskorski** [comment]
This error happens to me from time to time, but cannot reproduce this one. Under `sklearn==1.2.2` this seems working:
```
from sklearn import neighbors, datasets
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier

X,y = datasets.load_iris(return_X_y=True)

param_grid = [{'weights': ["uniform", "distance"], 'n_neighbors': [3, 4, 5]}]
knn_clf = KNeighborsClassifier()

grid_search = GridSearchCV(knn_clf, param_grid, cv=5, verbose=3, n_jobs=-1)
grid_search.fit(X,y)
```
